### PR TITLE
System integration tests for nanoDFT Demo Notebook

### LINKED ISSUE #38: Add github action for basic pass/fail testing of notebooks
Minimal CI requirements:

* Use [nbmake](https://pypi.org/project/nbmake/0.0.1/) 
* Runs on PR and main branch
* Using ipu model but could also have a separate task for running using the CPU backend.

### PR DESCRIPTION
**Pull request by hatemhelal**
Fixes #38 

This PR adds an initial workflow that uses the IPU model to check that
the nanoDFT demo notebook doesn't explode.

Some of the changes were necessary to ensure that everything runs with a
configurable number of tiles as well as working around the lack of
executable caching available on the IPU model.

This PR also made some changes to try and support testing the dataset
generation notebook.  This notebook is still failing and will need to be
addressed in a later PR if we want to extend this sort of system level
testing.

### CODE REVIEW

File: electron_repulsion/direct.py [hunk 1]
```diff
@@ -40,7 +40,7 @@ def get_atom_string(atoms, locs):
     return atom_string, str
 
 def num_tiles():
-        return 1472
+        return int(os.getenv('JAX_IPU_MODEL_NUM_TILES', 1472))
```
Comment (balancap) [inline electron_repulsion/direct.py:43]: I believe it is cleaner here to use `jax` device interface, i.e. `jax.devices("ipu")[0].num_tiles`

File: requirements.txt [hunk 2]
```diff
@@ -8,6 +8,7 @@ seaborn
 tqdm
 pytest
 natsort
+nbmake
```
Comment (balancap) [inline requirements.txt:11]: We should have a separate `test_requirements.txt` for test only dependencies. It does not make much sense to pull `pytest`, `nbmake`, ... for all users who just want to run a notebook/script

**Review by balancap — COMMENTED** [review]
Thanks @hatemhelal Just a couple of small comments :)

Did you check if the modified notebooks are still running on Paperspace? e.g.
```
import os
# Update working directory to root directory of the repo.
if len([d for d in os.listdir() if d.endswith(".ipynb")]) > 0:
    os.chdir(os.getcwd() + "/..")
print("Working directory:", os.getcwd())
```
that you changed has to do with setting up properly the root path in Paperspace (probably not the cleanest way!)

File: electron_repulsion/direct.py (hunk 1 above)
Comment (hatemhelal) [reply electron_repulsion/direct.py:43]: good point.

**Comment by hatemhelal** [comment]
> Did you check if the modified notebooks are still running on Paperspace? e.g.

How about I revert these changes to the data generator notebook?  This was pure speculation as I never got it to work but I thought the action runner was failing to find the downloaded files due to the working directory + relative paths.

File: requirements.txt (hunk 2 above)
Comment (hatemhelal) [reply requirements.txt:11]: this is a good idea, thanks!

**Comment by balancap** [comment]
@hatemhelal I pushed the fix on TessellateIPU repo and re-triggered the CI. Let's see :)

**Review by balancap — COMMENTED** [review]
CI passing :tada:

**Comment by balancap** [comment]
Rule of thumb on pinning dependencies: we can pin external ones (we don't have control on these), but not internal ones. For internal libraries, there is value at just trying to constantly use latest/main: getting last improved features, and helping quickly catching+solving bugs.

**Pull request merged**